### DISCUSSION TITLE
Implicit and repeated nodes

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#16] [General]
About the terminology:
I prefer **"implicit node"** rather than **"missing node"**, if they are valid.

Implicit nodes can be complemented after parsing process, in the application layer.
For example, implicit nodes can be added just before visualization if they are needed:
https://github.com/g2glab/pg-formatter/blob/c88fc3dfac9809a286e709ab7f3c1c10ed26a54f/lib/for-blitz.js#L35-L38

**Comment by hchiba1** [comment]
Implicit nodes can also be detected at the end of parsing.
However, the process can be resource-intensive for large files, and its necessity depends on the application.
Thus, it might be better to keep the complement process separate from the parser.

**Comment by hchiba1** [comment]
Not sure if the spec should describe the following two expressions are equivalent.
```
1 -- 2
```

```
1 -- 2
1
2
```

**Comment by nichtich** [comment]
Yes, implicit nodes are a handy feature and should be supported. This is somehow connected to the question of repeated node identifiers. What's the graph of this document:

~~~
a -> b
a :foo
a :bar
~~~

1.  ~~~
    a
    b
    a -> b
    ~~~

2.  ~~~
    a :foo
    b
    a -> b
    ~~~

3.  ~~~
    a :bar
    b
    a -> b
    ~~~

4.  ~~~
    a :foo :bar
    b
    a -> b
    ~~~

I'd say 3 (last node overrides preceding node with same identifier) or 4 (nodes of same identfier are merged).

**Comment by hchiba1** [comment]
I prefer 4, because
* those repeated lines do not conflict, so both can be retained without dropping information
* such behavior reminds me of multivalued properties, i.e., `k:v1 k:v2`means `k:v1,v2`
* shuffling lines of a file does not change the result (which seems to be a nice behavior)
* similarly, concatenating multiple PG files produces stable results, regardless of the file order
* finally, such features are similar to those of RDF (especially Turtle)

**Reply by nichtich** [reply]
Ok, I'll implement it this way.

**Comment by hchiba1** [comment]
In the case that the target format of a conversion does not allow for multiple labels, choosing option 3 seems reasonable to me.

**Comment by nichtich** [comment]
The discussion was about implicit and repeated nodes in PG format only.

- For PG-JSON I'd say implicit nodes MAY be allowed but repeated node identifiers are errors.
- For PG-JSONL I would allow implicit nodes but neither allow merging nodes nor make repeated node identifiers an error: it depends on the use case. PG-JSONL can be streamed for large graphs and this may or may not include overwriting a complete node.

**Reply by hchiba1** [reply]
I think this issue is somewhat related to the assumption of open world or closed world,
such that open world should allow incomplete data (implicit nodes) and accept additional information (repeated nodes).

Given that PG files can be concatenated,
a PG file might be more like an open world.
In this sense, it is reasonable for a PG file to allow implicit and repeated nodes.

In the same sense, the line-oriented PG-JSONL is very similar.
So I would like PG-JSONL to allow incomplete and repeated nodes.
(Thus, the PG-JSONL results in the same number of lines as PG, which is also reasonable.)

On the other hand, a PG-JSON file seems a bit different, representing an enclosed JSON object.
It is not a concatenable line-oriented format, and it looks more like a closed world.
It seems more natural that incomplete data (implicit nodes) and duplicates (repeated nodes) are not allowed.

**Reply by hchiba1** [reply]
(That said, there may be other cases where PG-JSON files represent incomplete data.)
One might argue that this issue should be separated from serialization formats.

**Reply by nichtich** [reply]
The [current specification of **PG-JSON**](https://github.com/pg-format/specification?tab=readme-ov-file#pg-json) does not allow implicit or repeated nodes.

The short specification of **PG-JSONL** in the same document contains the sentence

> Applications MAY accept objects that can automatically be transformed to valid form, for instance a missing `type` field inferred from existence of fields `from` and `to`.

so I'd assume this also covers implicit nodes but this could be made more clear and repeated nodes are not mentioned at all. I'd prefer repeated nodes to overwrite an existing node, so a PG-JSONL stream can also be used to update graphs (apart from deletions, this would be another issue for extension).

For **PG format** we already agreed to allow both implicit and repeated nodes (nodes of same identfier are merged).

**Reply by hchiba1** [reply]
Remaining question: Are the repeated nodes in PG-JSONL intended to overwrite existing data? 

This might be determined by the application. 
Perhaps, we could allow repeated nodes in the format without defining their specific behaviors.

**Reply by nichtich** [reply]
Good idea! I've added this paragraph to specification of PG-JSONL, what do you think?:

> Node objects SHOULD be given before their node identifiers are referenced in an edge object but applications MAY also create implicit node objects for this cases. Applications MAY allow multiple node objects with same node identifier in PG-JSONL but they MUST make clear whether nodes with repeated identifier are ignored, merged into existing nodes, or replace existing nodes.

**Reply by hchiba1** [reply]
Thanks!
